perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/modules/BBMRI.java

```diff
@@ -0,0 +1,161 @@
+		Member member = perun.getMembersManager().getMemberByUser(session, vo, user);
+
+		// get the field of application with the collections
+		Set<String> collectionIDsInApplication = getCollectionIDsFromApplication(session, app);
+
+		// get map of collection IDs to group from Perun
+		Map<String, Group> collectionIDsToGroupsMap = new HashMap<>();
+		Group collectionsGroup = perun.getGroupsManager().getGroupByName(session, vo, "collections");
+		for (Group group : perun.getGroupsManager().getSubGroups(session, collectionsGroup)) {
+			for (Group subgroup : perun.getGroupsManager().getSubGroups(session, group)) {
+				if ("representatives".equals(subgroup.getShortName())) {
+					Attribute collectionID = perun.getAttributesManager().getAttribute(session, subgroup, "urn:perun:group:attribute-def:def:collectionID");
+					collectionIDsToGroupsMap.put((String) collectionID.getValue(), subgroup);
+				}
+			}
+		}
+
+		// add user to all groups from the field on application
+		for (String collectionID : collectionIDsInApplication) {
+			perun.getGroupsManager().addMember(session, collectionIDsToGroupsMap.get(collectionID), member);
+		}
+
+		return app;
+	}
+
+	@Override
+	public Application rejectApplication(PerunSession session, Application app, String reason) throws PerunException {
+		return app;
+	}
+
+
+	/**
+	 * Checks whether all collection IDs found in user input really exists in Perun.
+	 * If not, CantBeApproved exception is thrown.
+	 *
+	 * @param session who approves the application
+	 * @param app     application
+	 * @return application
+	 * @throws CantBeApprovedException if at least one collection ID does not exist in Perun
+	 */
+	public Application beforeApprove(PerunSession session, Application app) throws PerunException {
+		// get perun and beans from session
+		Perun perun = session.getPerun();
+		Vo vo = app.getVo();
+
+		// get all collection IDs from Perun
+		Set<String> collectionIDsInPerun = new HashSet<>();
+		Group collectionsGroup = perun.getGroupsManager().getGroupByName(session, vo, "collections");
+		for (Group group : perun.getGroupsManager().getSubGroups(session, collectionsGroup)) {
+			for (Group subgroup : perun.getGroupsManager().getSubGroups(session, group)) {
+				if ("representatives".equals(subgroup.getShortName())) {
+					Attribute collectionID = perun.getAttributesManager().getAttribute(session, subgroup, "urn:perun:group:attribute-def:def:collectionID");
+					collectionIDsInPerun.add((String) collectionID.getValue());
+				}
+			}
+		}
+
+		// get the field of application with the collections
+		Set<String> collectionIDsInApplication = getCollectionIDsFromApplication(session, app);
+
+		// get non-existing collections
+		collectionIDsInApplication.removeAll(collectionIDsInPerun);
+
+		// difference must be empty
+		if (!collectionIDsInApplication.isEmpty()) {
+			throw new CantBeApprovedException("Collections " + collectionIDsInApplication + " does not exists.", "", "", "", true);
```

Comment:
I checked and if you don't specify ``reason`` param in ``CantBeApprovedException`` as ``"NOT_ACADEMIC"`` it will server default behavior in GUI - printing your exception message and allowing to "approve anyway" if it's specified as soft exception. Your code seems to be correct.